rpc/personal_api.go

```diff
@@ -1,30 +1,181 @@
 package rpc
 
 import (
+	"bytes"
 	"context"
+	"fmt"
+	"log"
+	"os"
+	"sync"
+	"time"
 
 	sdkcontext "github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	emintcrypto "github.com/cosmos/ethermint/crypto"
+	params "github.com/cosmos/ethermint/rpc/args"
+	"github.com/spf13/viper"
 
+	"github.com/ethereum/go-ethereum/accounts"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
 )
 
 // PersonalEthAPI is the eth_ prefixed set of APIs in the Web3 JSON-RPC spec.
 type PersonalEthAPI struct {
-	cliCtx    sdkcontext.CLIContext
-	nonceLock *AddrLocker
+	cliCtx      sdkcontext.CLIContext
+	ethAPI      *PublicEthAPI
+	nonceLock   *AddrLocker
+	keys        []emintcrypto.PrivKeySecp256k1
+	keyInfos    []keyring.Info
+	keybaseLock sync.Mutex
 }
 
 // NewPersonalEthAPI creates an instance of the public ETH Web3 API.
-func NewPersonalEthAPI(cliCtx sdkcontext.CLIContext, nonceLock *AddrLocker) *PersonalEthAPI {
-	return &PersonalEthAPI{
+func NewPersonalEthAPI(cliCtx sdkcontext.CLIContext, ethAPI *PublicEthAPI, nonceLock *AddrLocker, keys []emintcrypto.PrivKeySecp256k1) *PersonalEthAPI {
+	api := &PersonalEthAPI{
 		cliCtx:    cliCtx,
+		ethAPI:    ethAPI,
 		nonceLock: nonceLock,
+		keys:      keys,
 	}
+
+	infos, err := api.getKeybaseInfo()
+	if err != nil {
+		return api
+	}
+
+	api.keyInfos = infos
+	return api
+}
+
+func (e *PersonalEthAPI) getKeybaseInfo() ([]keyring.Info, error) {
+	e.keybaseLock.Lock()
+	defer e.keybaseLock.Unlock()
+
+	if e.cliCtx.Keybase == nil {
+		keybase, err := keyring.NewKeyring(
+			sdk.KeyringServiceName(),
+			viper.GetString(flags.FlagKeyringBackend),
+			viper.GetString(flags.FlagHome),
+			e.cliCtx.Input,
+			emintcrypto.EthSecp256k1Options()...,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		e.cliCtx.Keybase = keybase
+	}
+
+	return e.cliCtx.Keybase.List()
+}
+
+// ImportRawKey stores the given hex encoded ECDSA key into the key directory,
+// encrypting it with the passphrase.
+func (e *PersonalEthAPI) ImportRawKey(privkey, password string) (common.Address, error) {
+	_, err := crypto.HexToECDSA(privkey)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	return common.Address{}, nil
+}
+
+// ListAccounts will return a list of addresses for accounts this node manages.
+func (e *PersonalEthAPI) ListAccounts() ([]common.Address, error) {
+	addrs := []common.Address{}
+	for _, info := range e.keyInfos {
+		addressBytes := info.GetPubKey().Address().Bytes()
+		addrs = append(addrs, common.BytesToAddress(addressBytes))
+	}
+
+	return addrs, nil
+}
+
+// LockAccount will lock the account associated with the given address when it's unlocked.
```

Comment:
update comment saying that it copies the key?

Reply:
it doesn't copy the key, it deletes it from the `PersonalEthAPI`'s local keys, I can update it to say that